### PR TITLE
[X86] combineKSHIFT - fold kshiftr(kshiftr/extract_subvector(X,C1),C2) --> kshiftr(X,C1+C2)

### PR DESCRIPTION
**Pull request by RKSimon**
Merge serial KSHIFTR nodes, possibly separated by EXTRACT_SUBVECTOR, to allow mask instructions to be computed in parallel.

### CODE REVIEW

File: llvm/lib/Target/X86/X86ISelLowering.cpp [hunk 1]
```diff
@@ -58405,11 +58405,30 @@ static SDValue combineEXTEND_VECTOR_INREG(SDNode *N, SelectionDAG &DAG,
...
   EVT VT = N->getValueType(0);
-
+  const TargetLowering &TLI = DAG.getTargetLoweringInfo();
   if (ISD::isBuildVectorAllZeros(N->getOperand(0).getNode()))
     return DAG.getConstant(0, SDLoc(N), VT);
 
-  const TargetLowering &TLI = DAG.getTargetLoweringInfo();
+  // Fold kshiftr(extract_subvector(X,C1),C2)
+  //  --> extract_subvector(kshiftr(X,C1+C2),0)
+  // Fold kshiftr(kshiftr(X,C1),C2) --> kshiftr(X,C1+C2)
```
Comment (phoebewang) [inline llvm/lib/Target/X86/X86ISelLowering.cpp:58414]: Does this also apply to kshiftl?

**Comment by phoebewang** [comment]
> Merge serial KSHIFTR nodes, possibly separated by EXTRACT_SUBVECTOR, to allow mask instructions to be computed in parallel.

I think parallel is not much important for mask operations, becasue they mainly serve for vector instructions. That says, the contiguous mask operations would not be a bottleneck in real workload.

File: llvm/lib/Target/X86/X86ISelLowering.cpp (hunk 1 above)
Comment (RKSimon) [reply llvm/lib/Target/X86/X86ISelLowering.cpp:58414]: `(kshiftl (kshiftl x, c1), c2)` might be useful - not sure if we need to support `kshiftl+insert_subvector/concat_subvectors` patterns or not though.

**Comment by RKSimon** [comment]
@phoebewang I'd still like get this in, do you think this will just be neutral or actually cause regressions?

**Review by phoebewang — APPROVED** [review]
> @phoebewang I'd still like get this in, do you think this will just be neutral or actually cause regressions?

I'm neutral to the patch. I don't test the performance, but assume the effect is minor. So land as you like :)

**Pull request merged**